docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,529 @@
+   * Evaluate code that was transformed by Vite.
+   * @param context Function context
+   * @param code Transformed code
+   * @param id ID that was used to fetch the module
+   */
+  evaluateModule(
+    context: ModuleRunnerContext,
+    code: string,
+    id: string,
+  ): Promise<any>
+  /**
+   * evaluate externalized module.
+   * @param file File URL to the external module
+   */
+  evaluateExternalModule(file: string): Promise<any>
+}
+```
+
+Vite exports `ESModulesEvaluator` that implements this interface by default. It uses `new AsyncFunction` to evaluate code, so if the code has inlined source map it should contain an [offset of 2 lines](https://tc39.es/ecma262/#sec-createdynamicfunction) to accommodate for new lines added. This is done automatically in the server SSR environment. If your runner implementation doesn't have this constraint, you should use `fetchModule` (exported from `vite`) directly.
+
+## HMRModuleRunnerConnection
+
+**Type Signature:**
+
+```ts
+export interface HMRModuleRunnerConnection {
+  /**
+   * Checked before sending messages to the client.
+   */
+  isReady(): boolean
+  /**
+   * Send message to the client.
+   */
+  send(message: string): void
+  /**
+   * Configure how HMR is handled when this connection triggers an update.
+   * This method expects that connection will start listening for HMR updates and call this callback when it's received.
+   */
+  onUpdate(callback: (payload: HMRPayload) => void): void
+}
+```
+
+This interface defines how HMR communication is established. Vite exports `ServerHMRConnector` from the main entry point to support HMR during Vite SSR. The `isReady` and `send` methods are usually called when the custom event is triggered (like, `import.meta.hot.send("my-event")`).
+
+`onUpdate` is called only once when the new runtime is initiated. It passed down a method that should be called when connection triggers the HMR event. The implementation depends on the type of connection (as an example, it can be `WebSocket`/`EventEmitter`/`MessageChannel`), but it usually looks something like this:
+
+```js
+function onUpdate(callback) {
+  this.connection.on('hmr', (event) => callback(event.data))
+}
+```
+
+The callback is queued and it will wait for the current update to be resolved before processing the next update. Unlike the browser implementation, HMR updates in Vite Runtime wait until all listeners (like, `vite:beforeUpdate`/`vite:beforeFullReload`) are finished before updating the modules.
+
+## Backward Compatibility
+
+The current Vite server API will be deprecated but keep working during the next major.
+
+- `server.transformRequest(url)` <br> -> `server.environment('browser').transformRequest(url)`
+- `server.transformRequest(url, { ssr: true })` <br> -> `server.environment('ssr').tranformRequest(url)`
+- `server.warmupRequest(url)` <br> -> `server.environment('browser').warmupRequest(url)`
+- `server.ssrLoadModule(url)` <br> -> `server.ssrModuleRunner.executeEntryPoint(url)`
```

Comment:
Shouldn't it be `server.environment("ssr").run(url)`?

Reply:
I thought that `run(url)` would not return a Module instance, as in other environments it isn't possible to do that, no? `run(url)` has `void` return to me.

Reply:
Ah, you are right